### DISCUSSION TITLE
Negative Field in TFI with RBM gives wrong energy

### DISCUSSION THREAD

**Discussion by skothegh** [netket/netket#772] [Q&A] [answered]
Aloha,
I have been playing with your tutorial code and when doing a scan over the field strength, I found that the groundstate energy and correlation functions give wrong results for h<0. For positive fields it usually finds the correct state fast and reliably.

Minimal Example:

```
import netket as nk

L = 6  

g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)

hi = nk.hilbert.Spin(s=1/2, N=g.n_nodes)

ma = nk.models.RBM(alpha=4, use_visible_bias=True, dtype=float)

sa = nk.sampler.MetropolisLocal(hi, n_chains=16, n_sweeps = L, reset_chains=False)

op = nk.optimizer.Sgd(learning_rate=0.1)

sr = nk.optimizer.SR(diag_shift=0.1)

ha = nk.operator.Ising(hilbert=hi, graph=g, h=-1)

gs = nk.VMC(ha, op, sa, ma, sr = sr, n_samples=1000)
gs.run(n_iter=100, out='RBM')

evals = nk.exact.lanczos_ed(ha, compute_eigenvectors=False)
print(evals[0])
```

**Comment by PhilipVinc** [comment, marked as answer]
That's to be expected. 

The easy way out is to just use a complex-valued neural network because those are somewhat less sensible to the initial parameters, at least in this case, and can encode the sign.

Else you need to play with the initial distribution of the parameters. Probably widen the distribution a bit IIRC.
You also must be careful about the step of the SGD. That's quite large and might not always work well.

Also, 100 steps is very little.
We know that's enough for the example but if you vary the parameters (especially if you considerably increase the number of parameters, as you did here) you shouldn't expect this to be enough.

**Reply by gcarleo** [reply]
The main point is that the wave function can be negative if you change sign to the field, so there is no reason to expect that a real valued RBM (notice you are doing ```dtype=float```)
```python
ma = nk.models.RBM(alpha=4, use_visible_bias=True, dtype=float)
```

can capture it, since those states are positive definite. You need to work with wave functions that can change signs or phase...

**Reply by skothegh** [reply]
Aye, thanks. I completely overlooked that parameter. I just read your paper up and down and was sure that RBMs used as NNQS are always complex. Thanks a bunch for your help!
